### ISSUE TITLE
Histogram generation

### ISSUE THREAD

**Issue by z0u** [vpac-innovations/rsa#3]
rsaquery now has the ability to gather statistics for its output (see #1). Another filter should be added to gather histogram data. It should perform two tasks:
1. Split data into a natural histogram.
2. Split data based on some categorical metadata - for example, given an input band of data and an input metadata band of regions, it should produce one natural histogram per region.

It might be tricky to create a histogram without knowing the min, max, mean and standard deviation in advance. This is especially tricky because tiles might be processed on different nodes, and new time slices might be added at a later date which would invalidate any pre-computed histograms. To solve this, we could:
- Require histogram generation to be run in a separate query, with the basic stats as input. This would give a better guarantee of useful results, but would add to development and processing time.
- Use a common set of buckets for all data. One nice property of this is that, because all possible buckets are known in advance, buckets that have nothing in them don't need to be stored. Example bucket arrangements:
  - There could be a number of sets of uniformly-sized buckets, e.g. 0-1000 in buckets of 10, 0-10,000 in buckets of 100, etc. Note that these sets overlap.
  - We could use buckets that increase in size the further you get from zero, perhaps using a Fibonacci-like sequence e.g. 0, 0.1, 0.2, 0.3, 0.5, 0.8, 1.3, 2.1, etc. This allows for similar resolution regardless of the scale of the data - except where data is closely clustered at some high range. Negative numbers should be supported too.

**Comment by z0u** [comment]
Through a bit of trial and error in a spreadsheet, I've found that a nice sequence is generated by this equation:

![text2989](https://cloud.githubusercontent.com/assets/1408875/2664698/bf91da04-c07f-11e3-8223-ea8b045b5499.png)

> s \* (10 ^ (i / b))

Where _i_ is the bucket index, _b_ is the number of buckets per order of magnitude, and _s_ is the scaling factor. For example, with _s=0.1_ and _b=3_, the resulting bucket sequence is:

| i | lower bound |
| --- | --- |
| 0 | **0.1** |
| 1 | 0.215 |
| 2 | 0.464 |
| 3 | **1** |
| 4 | 2.15 |
| 5 | 4.64 |
| 6 | **10** |
| 7 | 21.5 |
| 8 | 46.4 |
| 9 | **100** |
| 10 | 215 |
| 11 | 464 |
| 12 | **1,000** |
| 13 | 2,150 |
| 14 | 4,640 |
| 15 | **10,000** |
| 16 | 21,500 |
| 17 | 46,400 |
| 18 | **100,000** |
| 19 | 215,000 |
| 20 | 464,000 |
| 21 | **1,000,000** |

Thus all positive values up to 1 million can be represented in 21 buckets.

To find the index, apply the inverse function:

![text2989](https://cloud.githubusercontent.com/assets/1408875/2664648/735a5c34-c07e-11e3-86fa-263538b9cb2c.png)

> (int) (b \* log10(x/s))

However, we don't want to have to compute a logarithm for every value. Instead, we should probably search through a pre-computed array.

To summarise, this is what we should do:
- Create a new class called `Hist` to complement `Stats`.
- In its constructor, `Hist` should:
  - Create an array of all the lower-bounds using the formula given above. This should be specially computed to allow negative numbers too. The array should accommodate reasonably large values, e.g. _-1e6_ - _1e6_.
  - Create an array of `Stats` objects with the same number of elements.
- During execution:
  - For each value read from the input, search through the lower bound array to find the index of the bucket.
  - Update the statistics being stored in that bucket with the current value.

**Comment by z0u** [comment]
For categorical bar charts (e.g. region-based), we should:
- Create a new class called `Cat`.
- In its constructor, `Cat` should create a hash map from `Int -> Hist`.
- During execution:
  - A value should be read from the categorical metadata layer.
  - Get a `Hist` object for this value from the hash map - or create it if it doesn't exist yet.
  - Keep the most recently used `Hist` object in a field for use next time in case the value doesn't change, to avoid hitting the hash map again. This is an optimisation for categorical data that doesn't change much from one pixel to the next - e.g. LGA regions.
  - Read a value from the real data layer and use it to update the `Hist` object.

**Comment by z0u** [comment]
`Cats` class implemented in 49d7837b7e125710b559c28a368289e0de4c48c7.

One last task: implement unit tests for `Cats`.

**Comment by z0u** [comment]
Done in 5ae1c8f2f003544e3efe5444b84045f825683901.

**Comment by z0u** [comment]
Just a quick note on speed: A benchmarking test has been added for sorting things into categories in 4a6df1118125bf94960bb430b8aad367906ec128. The test pushes 1000 categories and 25,000,000 values into a `Cats` object. This is the same number of values as there would be in a 5000x5000 tile. The test takes 1.35 seconds to run on my machine (average of 10 runs, with 2 warmup rounds).

[Victoria has an area of 237,629 km²](http://en.wikipedia.org/wiki/Victoria_%28Australia%29), so we need about 400,000,000 pixels to cover it; or 800,000,000 if you assume 50% wastage due to poor tile placement. This means we need 32 tiles, which would take 32_1.35 = 43 seconds to process on a single core. Note that this *_does not include IO**, so when working with real data this will be slower. However, these tests show that the current categories code should perform well enough for our needs, assuming IO is fast enough.

**Comment by z0u** [comment]
Actually it seems that there are about 56 tiles for a Vic-wide dataset, which suggests it would take 76 seconds of processor time.
